sys/net/gnrc/network_layer/ipv6/nib/_nib-arsm.c

```diff
@@ -519,7 +519,7 @@ static inline uint32_t _exp_backoff_retrans_timer(uint8_t ns_sent,
 
     /* backoff according to  https://tools.ietf.org/html/rfc7048 with
      * BACKOFF_MULTIPLE == 2 */
-    tmp = ((1 << ns_sent) * retrans_timer * tmp) / US_PER_MS;
+    tmp = ((((uint32_t) 1) << ns_sent) * retrans_timer * tmp) / US_PER_MS;
```

Comment:
I still have a question here, what value can `retrans_timer` can have. Because if it has a value greater or equal than 6, it could overflow with an `uint32_t` too.
If that is the case I would also cast after to an `uint64_t` like
`(uint64_t)((uint32_t)1 << ns_sent)`.

I did `((2 ** 32) -1) / ((2 ** 19) - 1) * 1500) == 5.4613437487483` to get the limit as I think `NDP_MAX_RANDOM_FACTOR` can be `1500` and `ns_sent` be `NDP_MAX_NS_NUMOF + 2` so `17 + 2`.

Reply:
`retrans_timer` can be way bigger than 6 but I don't really understand your calculation for the 6...

Reply:
But it is then divided by 1000 again.

Reply:
It would overflow before the division.

Reply:
Fine. Addressed.

Reply:
IMO there's no need to shift an `uint32_t` if it's casted to 64 bits anyway ..
```c
    uint32_t res = (uint32_t)((((uint64_t) 1) << ns_sent) * retrans_timer * factor) / US_PER_MS);
```
would do quite as good ..

Reply:
@cgundogan can you please confirm on an 8-bit platform. I don't have one at hand at the moment.

Reply:
yep, will give it a shot later today during h'n'a

Reply:
Thanks! (You'll probably have to take some tests out of the suite for that though ^^")

Reply:
At first sight I would be thinking that maybe doing a `uint32_t` shift would cost less code.

But it looks like, for a `arduino-atmega2560` that it reduces code size to use `uint64_t` all the way.

```
RIOT_CI_BUILD=1 CFLAGS=-DGNRC_IPV6_NIB_CONF_ARSM=1 USEMODULE=gnrc_ipv6_default BOARD=arduino-mega2560 make -C examples/gnrc_minimal clean all
make: Entering directory '/home/harter/work/git/RIOT/examples/gnrc_minimal'
Building application "gnrc_minimal" for "arduino-mega2560" with MCU "atmega2560".

   text    data     bss     dec     hex filename
  35820     514    2352   38686    971e /home/harter/work/git/RIOT/examples/gnrc_minimal/bin/arduino-mega2560/gnrc_minimal.elf
make: Leaving directory '/home/harter/work/git/RIOT/examples/gnrc_minimal'

# with uint64_t
RIOT_CI_BUILD=1 CFLAGS=-DGNRC_IPV6_NIB_CONF_ARSM=1 USEMODULE=gnrc_ipv6_default BOARD=arduino-mega2560 make -C examples/gnrc_minimal clean all
make: Entering directory '/home/harter/work/git/RIOT/examples/gnrc_minimal'
Building application "gnrc_minimal" for "arduino-mega2560" with MCU "atmega2560".

   text    data     bss     dec     hex filename
  35802     514    2352   38668    970c /home/harter/work/git/RIOT/examples/gnrc_minimal/bin/arduino-mega2560/gnrc_minimal.elf
make: Leaving directory '/home/harter/work/git/RIOT/examples/gnrc_minimal'
```

But on `samr21-xpro` it produces less code to do a `uint32_t` shift:

```
RIOT_CI_BUILD=1 CFLAGS=-DGNRC_IPV6_NIB_CONF_ARSM=1 USEMODULE=gnrc_ipv6_default BOARD=samr21-xpro make -C examples/gnrc_minimal clean all
make: Entering directory '/home/harter/work/git/RIOT/examples/gnrc_minimal'
Building application "gnrc_minimal" for "samr21-xpro" with MCU "samd21".

   text    data     bss     dec     hex filename
  43384     156    7828   51368    c8a8 /home/harter/work/git/RIOT/examples/gnrc_minimal/bin/samr21-xpro/gnrc_minimal.elf
make: Leaving directory '/home/harter/work/git/RIOT/examples/gnrc_minimal'

# with uint64_t
 RIOT_CI_BUILD=1 CFLAGS=-DGNRC_IPV6_NIB_CONF_ARSM=1 USEMODULE=gnrc_ipv6_default BOARD=samr21-xpro make -C examples/gnrc_minimal clean all
make: Entering directory '/home/harter/work/git/RIOT/examples/gnrc_minimal'
Building application "gnrc_minimal" for "samr21-xpro" with MCU "samd21".

   text    data     bss     dec     hex filename
  43400     156    7828   51384    c8b8 /home/harter/work/git/RIOT/examples/gnrc_minimal/bin/samr21-xpro/gnrc_minimal.elf
make: Leaving directory '/home/harter/work/git/RIOT/examples/gnrc_minimal'
```

Reply:
So how do I proceed :-)?

Reply:
18 bytes vs 16 bytes .. tough call (:

Reply:
The needed code is to the `<< ns_sent` on an `uint32_t` only. There is not technical reason to do it as `uint64_t` from this point. So I would keep the version with twe casts as it is now.